sycl/doc/extensions/supported/sycl_ext_oneapi_backend_level_zero.md

```diff
@@ -210,6 +210,59 @@ struct {
 
 [^1]: The SYCL implementation is responsible for distinguishing between the variants of <code>backend_input_t<backend::ext_oneapi_level_zero, queue></code>.
 
+### 4.1.1 Additional Experimental Mapping of SYCL objects to Level-Zero handles
+
+In support of SYCL queues constructed using either Level Zero commandqueue handles or commandlist handles
+the following interoperability types are supported:
+
+<table>
+<tr>
+<th>SyclType</th>
+<th>
+
+``` C++
+backend_return_t2<backend::ext_oneapi_level_zero,
+                 SyclType>
+```
+</th>
+<th>
+
+``` C++
+backend_input_t2<backend::ext_oneapi_level_zero,
+                SyclType>
+```
+</th>
+</tr><tr>
+<td rowspan="2">queue</td>
+<td rowspan="2">
+<pre>NativeHandleEnhanced_t</pre>
+</td>
+
+</tr><tr>
+<td>
+
+``` C++
+struct {
+  // ze_command_queue_handle_t or
+  // ze_command_list_handle_t
+  void *LevelZeroHandle;
+  bool IsImmCmdList;
+} NativeHandleEnhanced_t;
+
+struct {
+  NativeHandleEnhanced_t *NativeHandle;
+  device Device;
+  ext::oneapi::level_zero::ownership Ownership{
+      ext::oneapi::level_zero::ownership::transfer};
+}
+```
+
+Supported since version 4 of this specification.[^2]
+</td>
+
+</tr>
+</table>
+

```

Comment:
Adding some comments about the changes to the interop API here so the conversation can be threaded:

* I don't think it is necessary to add the `backend_input_t2` or `backend_return_t2` template types.  I think these just add unneeded complexity.  Can we instead just define two new free functions something like:

  ```
  namespace ext::oneapi::level_zero {

  queue make_queue(ze_command_list_handle_t native, const device &dev,
                   const context &ctxt, ownership own = ownership::transfer);

  std::variant<ze_command_list_handle_t, ze_command_queue_handle_t> get_native(const queue &queue);

  }
  ```

  We could even consider adding a second overload of `make_queue` to allow creating a queue either with an immediate command list or without:

  ```
  namespace ext::oneapi::level_zero {

  queue make_queue(ze_command_queue_handle_t native, const device &dev,
                   const context &ctxt, ownership own = ownership::transfer);

  queue make_queue(ze_command_list_handle_t native, const device &dev,
                   const context &ctxt, ownership own = ownership::transfer);

  std::variant<ze_command_list_handle_t, ze_command_queue_handle_t> get_native(const queue &queue);

  }
  ```

* The use of `void *LevelZeroHandle` in `NativeHandleEnhanced_t` is not modern C++.  Use `std::variant` like I show above.

* Do you want these new APIs to be experimental?  If so, add them to the namespace `ext::oneapi::level_zero::experimental`.  Also, I'd like to create a new extension specification for them to avoid mixing experimental and non-experimental APIs in the same extension.  This new extension specification should use the new [template](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc).

* What is the behavior of the existing `sycl::get_native(const queue &)` function when the queue has an immediate command list?

Reply:
+1 for using std::variant, but isn't standard requiring there is only one "make_queue" unless we'd be adding them as an extension: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:backend-interoperability-make ?

Reply:
The way it is implemented now, at the time of the next ABI change, we would simply replace the backend_input_t and backend_return_t for queue with the "t2" variants.

The make_queue variants you propose don't fit the existing make_queue (and all other make APIs) that take a single input parameter of type back_input_t. In your proposal the way make_queue is called differs considerably from the other make APIs.

Regarding your question about behavior of existing sycl::get_native(const queue &) function when the queue has an immediate command list, it will return a nullptr for the command_queue handle. That older API is not intended to be used with immediate command lists.

Reply:
Also, although I've used the word "experimental" in the PR, these changes are intended to be used today with immediate commandlists as standard practice. They are not being added to the spec immediately because they are an ABI-breaking changes.

Reply:
> but isn't standard requiring there is only one "make_queue" unless we'd be adding them as an extension

The spec does not prevent us from adding other functions to the `sycl::ext::level_zero` namespace, and we could name those functions `make_queue` and `get_native` if we want.  (Note that the spec versions of these functions are in the `sycl` namespace, so there is no collision.)  Therefore, the functions I show above are allowed by the core SYCL spec.

However, if our goal is to change the definition of `backend_input_t` and `backend_return_t` in the next ABI break period, there is some advantage to defining the `backend_input_t2` and `backend_return_t2` types as you have.  This way, code using the `t2` versions can more easily migrate after the ABI break.

It would be nice if we could avoid the need to support the `t2` versions forever.  Is it acceptable to the users of these new APIs if they are experimental until the next ABI break?  That would allow us to remove them without going through the deprecation period.  I *think* you are proposing that they are experimental, but the functions `get_native_standard_or_immediate` and `make_queue_standard_or_immediate` are not documented as such in this PR.

Reply:
> Regarding your question about behavior of existing sycl::get_native(const queue &) function when the queue has an immediate command list, it will return a nullptr for the command_queue handle.

The description of `sycl::get_native(const queue &)` should be updated to describe this behavior w.r.t. immediate command lists.

> That older API is not intended to be used with immediate command lists.

The application doesn't have any control over this, though.  I thought the runtime will construct a `queue` via an immediate command list without any changes to the application.  (I forget if this is the default now, or if it is enabled via an environment variable.  Either way, the application doesn't change.)

Technically, returning null in this case is an API break.  I don't see any way around this, though, other than delaying the immediate command list support until the next API/ABI break, and that seems like a bad option.